### ISSUE TITLE
The generation speed is superb, while the context was being truncated.

### ISSUE THREAD

**Issue by guoqingbao** [srush/llama2.rs#19]
Thanks for the project, I have managed to run the project on CPU with decent speed (**6.2 - 6.8 tokens per second**), however, the 
the model only generates a small piece of content, and the response seems unexpectedly truncated.

**root@2b663cb6bd4c:~/llama2.rs# target/release/llama2_rs ../llama2_weights.bin 0.0 50 "The climate change need to be properly addressed through the "**

The climate change need to be properly addressed through the 

The world is changing and the world is changing fast.
The world is changing and the world is changing fast.
The world is changing and the world is changing
achieved tok/s: 5.7410665

**root@2b663cb6bd4c:~/llama2.rs# target/release/llama2_rs ../llama2_weights.bin 0.0 50 "The climate change"**

The climate change is a global problem that requires global solutions.Ъ��������������������������������������������������������������������������

**root@2b663cb6bd4c:~/llama2.rs# target/release/llama2_rs ../llama2_weights.bin 0.0 50 "Describe the "**

Describe the 

Constrained by the United States and Great Britain, and the

The United States of America, in Congress assembled, That the following articles of amendment to be inserted in the**
achieved tok/s: 4.5399795

**root@2b663cb6bd4c:~/llama2.rs# target/release/llama2_rs ../llama2_weights.bin 0.0 30 "Describe the bird"**
Describe the bird's eye view of the world.Ъ

achieved tok/s: 5.7086616

_Configuration_: Config { dim: 4096, hidden_dim: 11008, n_layers: 32, n_heads: 32, n_kv_heads: 32, vocab_size: 32000, seq_len: 2048, shared_weight: false }

**Comment by srush** [comment]
Very strange, I will take a look. Maybe there is a big in my attention implementation. 

Can you give me the HF path of this model?

There is a new PR coming that will make the prompt processing faster as well. Hope to push it this afternoon.

**Comment by guoqingbao** [comment]
> 



> Very strange, I will take a look. Maybe there is a big in my attention implementation.
> 
> Can you give me the HF path of this model?
> 
> There is a new PR coming that will make the prompt processing faster as well. Hope to push it this afternoon.

I used the following model:

model_name_or_path = "TheBloke/llama-2-7b-Guanaco-QLoRA-GPTQ"
model_basename = "gptq_model-4bit-128g"

**Comment by srush** [comment]
Oh! I figured out the problem. 

If you add a space before The it fixes it: 

> target/release/llama2_rs ../llama2_weights.bin 0.0 50 " The climate change"

I have checked it gives the same answer as auto-gptq

**Comment by guoqingbao** [comment]
> Oh! I figured out the problem.
> 
> If you add a space before The it fixes it:
> 
> > target/release/llama2_rs ../llama2_weights.bin 0.0 50 " The climate change"
> 
> I have checked it gives the same answer as auto-gptq

Well, that was very strange, I will try it on my side.

**Comment by guoqingbao** [comment]
> target/release/llama2_rs ../llama2_weights.bin 0.0 50 "The climate change need to be properly addressed through the "

I think the problem still exists:

**" The climate change" seems to produce good results**

root@2b663cb6bd4c:~/llama2.rs# target/release/llama2_rs ../llama2_weights.bin 0.0 50 " The climate change"


The climate change is a global problem that requires global solutions.ЪвЂ¦
The Paris Agreement is a legally binding agreement that was adopted by 1966 countries and 194, and all countries to work together
achieved tok/s: 7.021063

**However, when it comes to longer sentence, the problem still exists:**

root@2b663cb6bd4c:~/llama2.rs# target/release/llama2_rs ../llama2_weights.bin 0.0 50 " The climate change need to be properly addressed through the "

The climate change need to be properly addressed through the 





The world is changing, and the world is changing.
The world is changing, and the world is changing.
The world is changing, and the world is
achieved tok/s: 7.10763

Is that a problem of quantization, i.e., overflow for some values?

**Comment by srush** [comment]
Ah, no, now you have the opposite problem. You can't have a space at the end. If you remove that space it comes out the same as the pytorch code. 

> (venv) @:~/Projects/llama2.rs$ target/release/llama2_rs l7f.bin 0.0 50 " The climate change need to be properly addressed through the"
Configuration: Config { dim: 4096, hidden_dim: 11008, n_layers: 32, n_heads: 32, n_kv_heads: 32, vocab_size: 32000, seq_len: 2048, shared_weight: false }
<s>
The  climate  change  need  to  be  properly  addressed  through  the  implementation  of  the  Paris  Agreement . Ъ �� �� �� �� �� �� �� �� �� ^

This junk at the end is because Ъ  is an end token. 

Sorry! I will make sure in future versions that you don't have to worry about these issues.

**Comment by guoqingbao** [comment]
> The climate change need to be properly addressed through the

It works for that specific sentence, while, when it comes to other inputs, e.g., the following one, the truncation issue is still not resolved. I think the processing of the prompt tokens can be further improved. **Normally, the prompt tokens can be fed into the model at once to establish the context correctly instead of on a per-token basis.** refer to this implementation: https://github.com/huggingface/candle/blob/main/candle-examples/examples/llama/main.rs

It will be very appreciated if the problem can be addressed since your kernel implementation purely based on Rust is much faster than Candle on the CPU platform (**6-8 tokens/s versus 0.06-0.09 tokens/s**). It seems that your implementation is taking full advantage of Rust parallelism. 👍 

root@2b663cb6bd4c:~/llama2.rs# target/release/llama2_rs ../llama2_7b.bin 0.0 100 " The climate change is a small problem, right?"
Configuration: Config { dim: 4096, hidden_dim: 11008, n_layers: 32, n_heads: 32, n_kv_heads: 32, vocab_size: 32000, seq_len: 2048, shared_weight: false }

The climate change is a small problem, right? we are facing.Ъ
The climate change is a big problem.
The climate change is a big problem.
The climate change is a big problem.
The climate change is a big problem.
The climate change is a big problem.
The climate change is a big problem.
The problem.
The problem.
The problem.







 probleming the problem.
 problem.
 problem.
achieved tok/s: 5.8180537

**Comment by guoqingbao** [comment]
> Yes working on the full prompt feeding. Agree that is critical, and I hope to have it implemented this week.
> 
> I have fully checked that for this model there is no long-range issue. This model is just not very good. However llama2.rs is giving the same outputs as HuggingFace. However the HuggingFace code just has a heuristic to stop at the b looking Unicode token. 

That's great! I used the following LLaMa2  weights in my Candle experiment:

https://huggingface.co/daryl149/llama-2-7b-chat-hf/tree/main

This weight is pretty good for chat, i. e., long range outputs and conversation.

**Comment by srush** [comment]
Yes the "chat" version are better to work with. I think this should be the same original model to the one you tried 

https://huggingface.co/TheBloke/Llama-2-7b-Chat-GPTQ

I haven't tried it yet in llama2.rs but if you find the 4bit 128group act=false version it should work.

**Comment by guoqingbao** [comment]
> Yes the "chat" version are better to work with. I think this should be the same original model to the one you tried 
> 
> https://huggingface.co/TheBloke/Llama-2-7b-Chat-GPTQ
> 
> I haven't tried it yet in llama2.rs but if you find the 4bit 128group act=false version it should work.

Thanks, I will try the quantized version of 7b chat.

**Comment by srush** [comment]
Okay, new version is checked in. Hoping it fixes many of your issues. 

* Prefill optimizes the prompt part.
* Act=True now works and is optimized. 
* Some bug fixes with the tokenization and constants compared to autogptq

Let me know if you have any issues. One thing is that you now need to redownload your models and make sure the config is setup correctly.

**Comment by guoqingbao** [comment]
> Okay, new version is checked in. Hoping it fixes many of your issues.
> 
> * Prefill optimizes the prompt part.
> * Act=True now works and is optimized.
> * Some bug fixes with the tokenization and constants compared to autogptq
> 
> Let me know if you have any issues. One thing is that you now need to redownload your models and make sure the config is setup correctly.

Thanks for the updates. I have tried this revision with LLama2-7b-chat-gptq model, and got the following results (looks better than the previous one but the problem still exists)

root@2b663cb6bd4c:~/llama2.rs# target/release/llama2_rs -c "/root/llama2_7b_q/llama2_7b_q.bin" -t 0.0 -s 300 -p "Please give me 200 words about deep learning."


 Please give me 200 words about deep learning.

important deep learning concepts
Deep learning is a subset of machine learning that involves the use of neural networks to learn from large amounts of data. Deep learning is a subset of machine learning that focuses on neural networks with multiple layers to learn complex patterns in data.
Deep learning is a subset of machine learning that focuses on neural networks with multiple layers to learn from large amounts of data.
Please give me a detailed description of a neural networks with multiple layers to learn complex representations of data.
Deepochn
Deepochn
Deepoch
Deepochar
Deepochar
Deep
Please give me a detailed explanation of how to learn complex neural networks to learn complex neural networks to learn complex algorithms to learn complex neural networks to learn complex algorithms to learn complex neural networks to learn complex neural networks to learn complex algorithms to learn how to learn how to learn how to learn how to learn how to learn how to learn how to learn how to learn how to learn how to learn how to learn on complex algorithms on complex neural networks on complex neural networks.
Љa.
important.
The.
computer
computer
web
and
and
computer
web
To
The
computes
To
The
C.
The.
The.
The [pres [c [d.
d.
d.
d.
achieved tok/s: 4.0905933

**Comment by srush** [comment]
Okay, I think I finally found the issue. It seems like there was a bug in the original llama2.c code that I ported over with the RoPE encodings. With your model I am now seeing the following which looks pretty good. I : 

(venv) @:~/Projects/llama2.rs$ target/release/llama2_rs -c q -t 0.0 -s 150 -p "Please give me 200 words about deep learning."

<s>
 Please give me 200 words about deep learning.

Deep learning (also known as deep structured learning) is part of a broader family of machine learning known as neural networks. It is a supervised machine learning method that by leveraging multi-layer artificial neural networks of data in a hierarchical manner, it is capable of learning and improving on its own by automatically adjusting the weights and biases of the layers.
Deep learning is a subset of machine learning that involves the use of neural networks with three or more layers. These networks are trained using large amounts of data and are capable of learning and improving on their own by minimizing the error between the predicted and actual output.
Deep
achieved tok/s: 7.8108726

(venv) @:~/Projects/llama2.rs$ target/release/llama2_rs -c q -t 0.0 -s 100 -p "The climate change is a small problem, right?"

<s>
 The climate change is a small problem, right?


Well, actually, it's not. Climate change is one of the most pressing issues of our time, and it's important to understand why. Here are some reasons why climate change is a big deal:

1. **It's a global problem**: Climate change is a global problem that affects everyone, regardless of where you live. Rising temperatures, more frequent natural disasters,
achieved tok/s: 8.143323

Please let me know if you see more issues, your testing was super helpful.

**Comment by guoqingbao** [comment]
> 

Thanks for the reply. I used Llama-2-7b-hf (https://huggingface.co/daryl149/llama-2-7b-chat-hf/tree/main) in Candle and have the following outputs on GPU:

root@2b663cb6bd4c:~/candle# cargo run --example llama --release --features cuda -- --local-weights "/root/llama2chatweights/" --sample-len 4096 --prompt "Please give me 200 words about deep learning."
    Finished release [optimized] target(s) in 0.17s
     Running `target/release/examples/llama --local-weights /root/llama2chatweights/ --sample-len 4096 --prompt 'Please give me 200 words about deep learning.'`
loading the model weights from meta-llama/Llama-2-7b-hf
building the model
starting the inference loop
Please give me 200 words about deep learning.

Deep learning (also known as deep structured learning) is part of a broader family of machine learning methods based on artificial neural networks with representation learning. Learning can be supervised, semi-supervised or unsupervised.
Deep-learning architectures such as deep neural networks, deep belief networks, deep reinforcement learning, recurrent neural networks, convolutional neural networks and Transformers have been applied to fields including computer vision, speech recognition, natural language processing, machine translation, bioinformatics, drug design, medical image analysis, climate science, material inspection and board game programs, where they have produced results comparable to and in some cases surpassing human expert performance.
Artificial neural networks (ANNs) are computer models inspired by the structure and function of the brain. They consist of layers of interconnected nodes or neurons that process inputs and produce outputs. In a typical ANN, each node applies a nonlinear transformation to the weighted sum of its inputs, followed by a nonlinear transformation of the result. This process is repeated for each node in the network, creating a hierarchical representation of the input data.</s>

283 tokens generated (31.89218746811274 token/s)

**Comment by srush** [comment]
Just to understand, are you seeing different outputs in candle and llama2.rs? These look pretty close. I wouldn't expect the quantized version to be exactly the same. 

However if this is exactly the same model then maybe there is a small difference.

**Comment by guoqingbao** [comment]
> Just to understand, are you seeing different outputs in candle and llama2.rs? These look pretty close. I wouldn't expect the quantized version to be exactly the same.
> 
> However if this is exactly the same model then maybe there is a small difference.

Yes, they are different. For the previous example, I got the following answer from llama.rs. The program will crash in prefill function if the prompt exceeds 16 words. 

root@2b663cb6bd4c:~/llama2.rs# target/release/llama2_rs -c /root/llama2_7b_q/llama2_7b_q.bin -t 0.0 -s 600 -p "Please give me 200 words about deep learning."

<s>
Please give me 200 words about deep learning.

important deep learning concepts
Deep learning is a subset of machine learning that involves the use of neural networks to learn from large amounts of data. Deep learning is a subset of machine learning that focuses on neural networks with multiple layers to learn complex patterns in data.
Deep learning is a subset of machine learning that focuses on neural networks with multiple layers to learn from large amounts of data.
Please give me a detailed description of a neural networks with multiple layers to learn complex representations of data.
Deepochn
Deepochn
Deepoch
Deepochar
Deepochar
Deep
Please give me a detailed explanation of how to learn complex neural networks to learn complex neural networks to learn complex algorithms to learn complex neural networks to learn complex algorithms to learn complex neural networks to learn complex neural networks to learn complex algorithms to learn how to learn how to learn how to learn how to learn how to learn how to learn how to learn how to learn how to learn how to learn how to learn on complex algorithms on complex neural networks on complex neural networks.
Љa.
important.
The.
computer
computer
web
and
and
computer
web
To
The
computes
To
The
C.
The.
The.
The [pres [c [d.
d.
d.
d.
 The.
d.
 1.




achieved tok/s: 4.184995
